### PR TITLE
Use wget instead of Batch's resource files for file downloads

### LINKED ISSUE #387: Azure Batch Resource Files occasionally timeout when downloading TES runner under load
**Describe the bug**
Batch occassionally times out when download the TES runner when under load.  It appears to be caused by insufficient retry logic.

**Steps to Reproduce**
Submit a large number of tasks at once

### PR DESCRIPTION
**Pull request by MattMcL4475**
Under load, the Batch service's code that downloads Resource Files seems to timeout and/or not be resilient.  This PR replaces all use of Batch Resource Files with wget.

Resolves #387

### CODE REVIEW

File: src/TesApi.Web/BatchScheduler.cs [hunk 1]
```diff
@@ -1363,14 +1388,16 @@ private async Task<StartTask> StartTaskIfNeeded(VirtualMachineConfiguration mach
             {
                 var commandLine = new StringBuilder();
                 commandLine.Append(@"/usr/bin/bash -c 'trap ""echo Error trapped; exit 0"" ERR; sudo touch tmp2.json && (sudo cp /etc/docker/daemon.json tmp1.json || sudo echo {} > tmp1.json) && sudo chmod a+w tmp?.json && if fgrep ""$(dirname ""$(dirname ""$AZ_BATCH_NODE_ROOT_DIR"")"")/docker"" tmp1.json; then echo grep ""found docker path""; elif [ $? -eq 1 ]; then ");
+
                 commandLine.Append(machineConfiguration.NodeAgentSkuId switch
                 {
                     var s when s.StartsWith("batch.node.ubuntu ") => "sudo apt-get install -y jq",
```
Comment (BMurri) [inline src/TesApi.Web/BatchScheduler.cs:1394]: You know that Ubuntu always has wget installed, or is this line missing wget?

File: src/TesApi.Web/BatchScheduler.cs [hunk 2]
```diff
@@ -1149,21 +1181,14 @@ private async Task<CloudTask> ConvertTesTaskToBatchTaskAsync(string taskId, TesT
...
-            var resourceFiles = new List<ResourceFile>
-            {
-                ResourceFile.FromUrl(await storageAccessProvider.GetInternalTesBlobUrlAsync(NodeTaskRunnerFilename, cancellationToken), NodeTaskRunnerFilename),
-                await UploadBlobAsync(NodeRunnerTaskInfoFilename, SerializeNodeTask(nodeTaskRunnerContent)),
-                await UploadBlobAsync(BatchScriptFileName, sb.ToString()),
-                await UploadBlobAsync(NodeTaskRunnerStartTaskFileName, SerializeNodeTask(uploadStartTaskFilesContent))
-            };
+            await storageAccessProvider.UploadBlobAsync(new Uri(nodeBatchScriptSasUrl), sb.ToString(), cancellationToken);
 
             var batchRunCommand = enableBatchAutopool
```
Comment (BMurri) [inline src/TesApi.Web/BatchScheduler.cs:1186]: I don't see where you download `nodeBatchScriptSasUrl` to `BatchScriptFileName` in the task command line before calling it.

**Review by BMurri — CHANGES_REQUESTED** [review]

File: src/TesApi.Web/BatchScheduler.cs [hunk 3]
```diff
@@ -1091,13 +1110,26 @@ private async Task<CloudTask> ConvertTesTaskToBatchTaskAsync(string taskId, TesT
...
             sb.AppendLinuxLine($"mkdir -p $AZ_BATCH_TASK_WORKING_DIR/wd && \\");
-            sb.AppendLinuxLine($"(./{NodeTaskRunnerFilename} upload --file {NodeTaskRunnerStartTaskFileName} || :) && \\"); // Upload the start-task console spews
+
+            var nodeTaskRunnerUrl = await storageAccessProvider.GetInternalTesBlobUrlAsync(NodeTaskRunnerFilename, cancellationToken);
+            var nodeTaskRunnerInfoUrl = await UploadBlobAsync(NodeRunnerTaskInfoFilename, SerializeNodeTask(nodeTaskRunnerContent));
+            var nodeBatchScriptSasUrl = await storageAccessProvider.GetInternalTesTaskBlobUrlAsync(task, BatchScriptFileName, cancellationToken);
+            var nodeRunnerStartTaskUrl = await UploadBlobAsync(NodeTaskRunnerTaskInfoToUploadStartTaskLogsFileName, SerializeNodeTask(startTaskStdOutStdErrUploadNodeTaskContent));
+            
+            sb.AppendLinuxLine($"write_ts DownloadRunnerScriptsStart && \\");
+            sb.AppendLinuxLine(CreateWgetCommand(nodeBatchScriptSasUrl, BatchScriptFileName) + " && \\");
```
Comment (BMurri) [inline src/TesApi.Web/BatchScheduler.cs:1126]: This line is trying to download this file content. This is what needs to be added to the command-line

File: src/TesApi.Web/BatchScheduler.cs (hunk 1 above)
Comment (MattMcL4475) [reply src/TesApi.Web/BatchScheduler.cs:1394]: yes Ubuntu always has wget installed

File: src/TesApi.Web/BatchScheduler.cs [hunk 4]
```diff
@@ -1149,21 +1180,14 @@ private async Task<CloudTask> ConvertTesTaskToBatchTaskAsync(string taskId, TesT
...
-                await UploadBlobAsync(BatchScriptFileName, sb.ToString()),
-                await UploadBlobAsync(NodeTaskRunnerStartTaskFileName, SerializeNodeTask(uploadStartTaskFilesContent))
-            };
+            await storageAccessProvider.UploadBlobAsync(new Uri(nodeBatchScriptSasUrl), sb.ToString(), cancellationToken);
 
             var batchRunCommand = enableBatchAutopool
-                ? $"/bin/bash -c chmod u+x ./{NodeTaskRunnerFilename} && /bin/bash $AZ_BATCH_TASK_WORKING_DIR/{BatchScriptFileName}"
-                : $"/bin/bash -c \"{MungeBatchScript()}\"";
+                ? $"/bin/bash -c {CreateWgetCommand(nodeBatchScriptSasUrl, BatchScriptFileName, setExecutable: true)} && ./{BatchScriptFileName}"
+                : $"/bin/bash -c {CreateWgetCommand(nodeBatchScriptSasUrl, BatchScriptFileName, setExecutable: true)} && \"{MungeBatchScript()}\"";
```
Comment (MattMcL4475) [inline src/TesApi.Web/BatchScheduler.cs:1187]: @BMurri  is this correct?

**Review by BMurri — APPROVED** [review]

File: src/TesApi.Web/BatchScheduler.cs [hunk 5]
```diff
@@ -1091,13 +1109,24 @@ private async Task<CloudTask> ConvertTesTaskToBatchTaskAsync(string taskId, TesT
 
             if (isPublic.DockerInDockerImage)
             {
-                sb.AppendLinuxLine($"(grep -q alpine /etc/os-release && apk add bash || :) && \\");  // Install bash if running on alpine (will be the case if running inside "docker" image)
+                sb.AppendLinuxLine($"(grep -q alpine /etc/os-release && apk add bash wget || :) && \\");  // Install bash if running on alpine (will be the case if running inside "docker" image)
```
Comment (BMurri) [inline src/TesApi.Web/BatchScheduler.cs:1112]: You probably will also want to add `wget` to [`microsoft/CromwellOnAzure/src/deploy-cromwell-on-azure/samples/docker-dockerfile`](https://github.com/microsoft/CromwellOnAzure/blob/main/src/deploy-cromwell-on-azure/samples/docker-dockerfile)

File: src/TesApi.Web/BatchScheduler.cs [hunk 6]
```diff
@@ -344,6 +344,25 @@ private async Task DeleteBatchTaskAndOrJobAndOrPoolIfExists(IAzureProxy azurePro
...
+        /// <summary>
+        /// Creates a wget command to robustly download a file
+        /// </summary>
+        /// <param name="url">URL to download</param>
+        /// <param name="localFilename">Filename for the output file</param>
+        /// <param name="setExecutable">Whether the file should be made executable or not</param>
+        /// <returns></returns>
+        private string CreateWgetCommand(string url, string localFilename, bool setExecutable = false)
+        {
+            string command = $"wget --https-only --timeout=20 --waitretry=1 --tries=5 --retry-connrefused --continue -O '{localFilename}' '{url}'";
```
Comment (BMurri) [inline src/TesApi.Web/BatchScheduler.cs:356]: Is this enough retries to change the way we download onto the compute nodes?

**Review by BMurri — COMMENTED** [review]
The integration test returned `Batch task ExitCode: 1`.

Could you take a current deployment, replace TES with this build, hand it a task, and look in batch while the task is queued waiting for the pool to add a compute node to see the exact command-line passed to the task?

File: src/TesApi.Web/BatchScheduler.cs (hunk 6 above)
Comment (MattMcL4475) [reply src/TesApi.Web/BatchScheduler.cs:356]: There are 5 retries; worst case scenario is 100 seconds of timeouts.  How many retries were you thinking?

Comment (BMurri) [reply src/TesApi.Web/BatchScheduler.cs:356]: The task runner retries 9 times before giving up.

**Pull request merged**